app/lib/oidc_client.rb

```diff
@@ -73,26 +75,50 @@ def set_checker_attribute(value:, access_token:, refresh_token:)
       access_token: access_token,
       refresh_token: refresh_token,
       method: :put,
+      uri: attribute_uri,
       arg: { value: value.to_json },
     )
   end
 
+  def has_email_subscription(access_token:, refresh_token:)
+    response = oauth_request(
+      access_token: access_token,
+      refresh_token: refresh_token,
+      method: :get,
+      uri: email_subscription_uri,
+    )
+
+    response.merge(result: (200..299).include?(response[:result].status))
+  end
+
+  def update_email_subscription(slug:, access_token:, refresh_token:)
+    oauth_request(
+      access_token: access_token,
+      refresh_token: refresh_token,
+      method: :post,
+      uri: email_subscription_uri,
+      arg: { topic_slug: slug },
+    )
+  end
+
 private
 
-  def oauth_request(access_token:, refresh_token:, method:, arg: nil)
+  OK_STATUSES = [200, 204, 404, 410].freeze
+
+  def oauth_request(access_token:, refresh_token:, method:, uri:, arg: nil)
     access_token_str = access_token
     refresh_token_str = refresh_token
 
-    args = [attribute_uri, arg].compact
+    args = [uri, arg].compact
 
     response = Rack::OAuth2::AccessToken::Bearer.new(access_token: access_token_str).public_send(method, *args)
 
-    unless [200, 404].include? response.status
+    unless OK_STATUSES.include? response.status
       client.refresh_token = refresh_token
       access_token = client.access_token!
 
       response = access_token.public_send(method, *args)
-      raise OAuthFailure unless [200, 404].include? response.status
+      raise OAuthFailure unless OK_STATUSES.include? response.status
```

Comment:
then if the refreshed token isn't 200, 204, 404, 410 then we raise an OAuthFailure.

Wouldn't we always want this to be 200?
When is 204, 404, 410 relevant?

Reply:
ah ha: https://github.com/alphagov/govuk-account-manager-prototype/pull/194#pullrequestreview-509327925

Reply:
The email endpoint can return a 204, 404, or 410 legitimately.  It could instead always return a 200 and say what the response really means in the body, but that feels like reinventing status codes.